{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate performance based on genetic distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the required packages\n",
    "import pandas as pd\n",
    "from itertools import product\n",
    "import sklearn.metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_bp_distances = [25, 50, 100, 500, 1000, 2000, \"depict\"]\n",
    "refsets = [\"farashi\", \"farashi p-value cutoff\", \"DeRycke\", \"Teslovich\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_metrics = pd.DataFrame(list(product(refsets, all_bp_distances)), columns = [\"refset\", \"bp distance\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "for am_index, am_values in all_metrics.iterrows():\n",
    "    # Load the reference set\n",
    "    if am_values[\"refset\"] == \"farashi\":\n",
    "        ref = pd.read_csv(\"/Users/vlietstraw/git/Post-GWAS/Input sets/Farashi/Farashi full 2000000 bp distance no pvalue filtering.csv\")\n",
    "\n",
    "    if am_values[\"refset\"] == \"farashi p-value cutoff\":\n",
    "        ref = pd.read_csv(\"/Users/vlietstraw/git/Post-GWAS/Input sets/Farashi/Farashi full 2000000 bp distance no pvalue filtering.csv\")\n",
    "        ref = ref[ref[\"GWAS/eQTL p-valueÂ¥\"] <= float(\"5e-8\")]\n",
    "\n",
    "    if am_values[\"refset\"] == \"DeRycke\":\n",
    "        ref = pd.read_csv(\"/Users/vlietstraw/git/Post-GWAS/Input sets/DeRycke/DeRycke reference set.csv\", delimiter = \";\")\n",
    "        ref.columns = [\"SNP ID\", \"chromosome\", \"location\", \"gene_ids\", \"gene name\", \"gene start\", \"gene stop\", \"Diff expression\", \"Class\", \"bp distance absolute\", \"bp distance\", \"Gene rank\"]\n",
    "\n",
    "    if am_values[\"refset\"] == \"Teslovich\":\n",
    "        ref = pd.read_csv(\"/Users/vlietstraw/git/Post-GWAS/Input sets/Teslovich/Teslovich reference set.csv\")\n",
    "        ref.columns = [\"SNP ID\", \"chromosome\", \"location\", \"P\", \"gene_ids\", \"gene name\", \"gene start\", \"gene stop\", \"Class\", \"bp distance absolute\", \"bp distance\", \"Gene rank\"]\n",
    "        \n",
    "    if am_values[\"bp distance\"] != \"depict\":\n",
    "        max_bp_distance = am_values[\"bp distance\"]\n",
    "        max_bp_distance = max_bp_distance * 1000\n",
    "        ref = ref[ref[\"bp distance absolute\"] <= max_bp_distance]\n",
    "    elif am_values[\"bp distance\"] == \"depict\":\n",
    "        if am_values[\"refset\"] == \"farashi\":\n",
    "            depict = pd.read_csv(\"~/git/DEPICT/outcomes/Farashi complete 2nd round/farashi_no_pvalue_filtering_geneprioritization.txt\", sep = \"\\t\")\n",
    "        if am_values[\"refset\"] == \"farashi p-value cutoff\":\n",
    "            depict = pd.read_csv(\"~/git/DEPICT/outcomes/Farashi complete 2nd round/farashi_default_pvalue_filtering_geneprioritization.txt\", sep = \"\\t\")\n",
    "        if am_values[\"refset\"] == \"DeRycke\":\n",
    "            depict = pd.read_csv(\"~/git/DEPICT/outcomes/DeRycke/DeRycke_output_geneprioritization.txt\", sep = \"\\t\")\n",
    "        if am_values[\"refset\"] == \"Teslovich\":\n",
    "            depict = pd.read_csv(\"~/git/DEPICT/outcomes/Teslovich for paper Wytze/Teslovich_output_geneprioritization.txt\", sep = \"\\t\")\n",
    "\n",
    "        depict[\"Locus\"] = depict[\"Locus\"].astype(str).apply(lambda x: x.split(\";\"))\n",
    "        depict = depict.explode(\"Locus\")\n",
    "\n",
    "        snp_replacement_dict = {\"rs113645266\" : \"rs6557271\",\n",
    "                        \"rs150282463\" : \"rs13137700\",\n",
    "                        \"rs67276543\" : \"rs34884832\"}\n",
    "        depict[\"Locus\"] = depict[\"Locus\"].replace(snp_replacement_dict)\n",
    "\n",
    "        depict = depict[[\"Locus\", \"Ensembl gene ID\"]]\n",
    "        depict.columns = [\"SNP ID\", \"gene_ids\"]\n",
    "\n",
    "        ref = ref.merge(depict, on = [\"SNP ID\", \"gene_ids\"], how = \"inner\")\n",
    "        \n",
    "        # Drop all unmappable candidates\n",
    "        ref.dropna(subset = [\"gene_ids\"], inplace = True)\n",
    "        \n",
    "        ref = ref.sort_values([\"SNP ID\", \"bp distance absolute\"], ascending = False)\n",
    "        ref[\"Gene rank\"] = ref.groupby(\"SNP ID\").cumcount() + 1\n",
    "    \n",
    "    # Drop all SNPs which no longer have a positive case\n",
    "    pos_counts = ref.groupby(\"SNP ID\")[\"Class\"].sum()\n",
    "    ref = ref[~ref[\"SNP ID\"].isin(pos_counts[pos_counts == 0].index)]\n",
    "    \n",
    "    all_metrics.at[am_index, \"Recall snps\"] = len(set(ref[\"SNP ID\"]))\n",
    "    all_metrics.at[am_index, \"Recall combinations\"] = sum(ref[\"Class\"])\n",
    "    \n",
    "    fpr, tpr, thresholds = sklearn.metrics.roc_curve(ref[\"Class\"], -ref[\"Gene rank\"], pos_label = 1)\n",
    "    all_metrics.at[am_index, \"ROC-AUC overall (lso)\"] = sklearn.metrics.auc(fpr, tpr) * 100\n",
    "\n",
    "    # Calculate the ROC-AUC for every SNP and average the result\n",
    "    SNPS2 = list(set(ref[\"SNP ID\"]))\n",
    "    aucs = []\n",
    "    for snp in SNPS2:\n",
    "      if len(set(ref[\"Class\"][ref[\"SNP ID\"] == snp])) == 1:\n",
    "          aucs.append(list(set(ref[\"Class\"][ref[\"SNP ID\"] == snp]))[0])\n",
    "      else:\n",
    "          fpr, tpr, thresholds = sklearn.metrics.roc_curve(ref[\"Class\"][ref[\"SNP ID\"] == snp], -ref[\"Gene rank\"][ref[\"SNP ID\"] == snp], pos_label = 1)\n",
    "          aucs.append(sklearn.metrics.auc(fpr, tpr))\n",
    "    all_metrics.at[am_index, \"ROC-AUC - mean per snpl (lso)\"] = sum(aucs)/len(aucs)\n",
    "\n",
    "\n",
    "    # In[22]:\n",
    "\n",
    "\n",
    "    # Calculate hits @1\n",
    "    all_metrics.at[am_index, \"Hits@1(lso)\"] = sum(ref[\"Class\"][(ref[\"Class\"] == 1) & (ref[\"Gene rank\"] == 1)])\n",
    "\n",
    "\n",
    "    # In[23]:\n",
    "\n",
    "\n",
    "    # Calculate hits @3\n",
    "    all_metrics.at[am_index, \"Hits@3(lso)\"] = sum(ref[\"Class\"][(ref[\"Class\"] == 1) & (ref[\"Gene rank\"] <= 3)])\n",
    "\n",
    "\n",
    "    # In[24]:\n",
    "\n",
    "\n",
    "    # Calculate hits @5\n",
    "    all_metrics.at[am_index, \"Hits@5(lso)\"] = sum(ref[\"Class\"][(ref[\"Class\"] == 1) & (ref[\"Gene rank\"] <= 5)])\n",
    "\n",
    "\n",
    "    # In[25]:\n",
    "\n",
    "\n",
    "    # Calculate hits @10\n",
    "    all_metrics.at[am_index, \"Hits@10(lso)\"] = sum(ref[\"Class\"][(ref[\"Class\"] == 1) & (ref[\"Gene rank\"] <= 10)])\n",
    "\n",
    "\n",
    "    # In[26]:\n",
    "\n",
    "\n",
    "    all_metrics.at[am_index, \"Mean rank (lso)\"] = ref[\"Gene rank\"][(ref[\"Class\"] == 1)].mean()\n",
    "\n",
    "\n",
    "    # In[27]:\n",
    "\n",
    "\n",
    "    all_metrics.at[am_index, \"Median rank (lso)\"] = ref[\"Gene rank\"][ref[\"Class\"] == 1].quantile(q = [0,0.25,0.5,0.75,1])[0.50]\n",
    "    \n",
    "    all_metrics.at[am_index, \"Mean number of candidates for every SNP\"] = ref.groupby(\"SNP ID\")[\"Gene rank\"].nunique().mean()\n",
    "    all_metrics.at[am_index, \"Mean number of negative candidates for every SNP\"] = ref[ref[\"Class\"] == 0].groupby(\"SNP ID\")[\"Gene rank\"].nunique().mean()\n",
    "    all_metrics.at[am_index, \"Mean number of positive candidates for every SNP\"] = ref[ref[\"Class\"] == 1].groupby(\"SNP ID\")[\"Gene rank\"].nunique().mean()\n",
    "    all_metrics.at[am_index, \"Median number of candidates for every SNP\"] = ref.groupby(\"SNP ID\")[\"Gene rank\"].nunique().quantile(q = [0,0.25,0.5,0.75,1])[0.50]\n",
    "    all_metrics.at[am_index, \"Median number of negative candidates for every SNP\"] = ref[ref[\"Class\"] == 0].groupby(\"SNP ID\")[\"Gene rank\"].nunique().quantile(q = [0,0.25,0.5,0.75,1])[0.50]\n",
    "    all_metrics.at[am_index, \"Min number of candidates for every SNP\"] = ref.groupby(\"SNP ID\")[\"Gene rank\"].nunique().min()\n",
    "    all_metrics.at[am_index, \"Max number of candidates for every SNP\"] = ref.groupby(\"SNP ID\")[\"Gene rank\"].nunique().max()\n",
    "    all_metrics.at[am_index, \"Min number of negative candidates for every SNP\"] = ref[ref[\"Class\"] == 0].groupby(\"SNP ID\")[\"Gene rank\"].nunique().min()\n",
    "    all_metrics.at[am_index, \"Max number of negative candidates for every SNP\"] = ref[ref[\"Class\"] == 0].groupby(\"SNP ID\")[\"Gene rank\"].nunique().max()\n",
    "    all_metrics.at[am_index, \"Total number of negative genes\"] = len(set(ref[ref[\"Class\"] == 0][\"gene_ids\"]))\n",
    "    all_metrics.at[am_index, \"Total number of positive genes\"] = len(set(ref[ref[\"Class\"] == 1][\"gene_ids\"]))\n",
    "    all_metrics.at[am_index, \"Total number of negative cases\"] = len(ref[ref[\"Class\"] == 0])\n",
    "    all_metrics.at[am_index, \"Total number of cases\"] = len(ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_metrics.to_csv(\"/Users/vlietstraw/git/Post-GWAS/all_variations_performance_metrics based on genetic distance.csv\", sep = \";\", decimal = \",\", index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
